apps/ae_socket_connector/lib/session_holder.ex

```diff
@@ -60,38 +84,104 @@ defmodule SessionHolder do
   end
 
   # Server
-  def init({%SocketConnector{} = socket_connector_state, ae_url, network_id, priv_key, color}) do
-    {:ok, pid} = SocketConnector.start_link(socket_connector_state, ae_url, network_id, color, self())
+  def init({socket_connector_state, ae_url, network_id, priv_key, connection_callbacks, file_name, :open, color}) do
+    :dets.open_file(String.to_atom(file_name), [type: :duplicate_bag])
+    {:ok, pid} = SocketConnector.start_link(socket_connector_state, ae_url, network_id, connection_callbacks, color, self())
 
     {:ok,
      %__MODULE__{
        socket_connector_pid: pid,
        socket_connector_state: socket_connector_state,
        network_id: network_id,
        priv_key: priv_key,
-       color: color
+       connection_callbacks: connection_callbacks,
+       color: color,
+       file: file_name
      }}
   end
 
+  def init({socket_connector_state, _ae_url, network_id, priv_key, connection_callbacks, file_name, :reestablish, color}) do
+    :dets.open_file(String.to_atom(file_name), [type: :duplicate_bag])
+    state =
+      %__MODULE__{
+        # socket_connector_pid: pid,
+        socket_connector_state: socket_connector_state,
+        network_id: network_id,
+        priv_key: priv_key,
+        connection_callbacks: connection_callbacks,
+        color: color,
+        file: file_name
+        }
+    {pid, saved_socket_connector_state} = reestablish_(state)
+    {:ok, %__MODULE__{state | socket_connector_state: Map.merge(saved_socket_connector_state, socket_connector_state), socket_connector_pid: pid, connection_callbacks: connection_callbacks}}
+  end
+
   defp kill_connection(pid, color) do
     Logger.debug("killing connector #{inspect(pid)}", ansi_color: color)
     Process.exit(pid, :normal)
   end
 
-  def fetch_state(pid) do
-    SocketConnector.request_state(pid)
+  # this is persent as a helper if you loose your channel id or password, then check this file.
+  defp persist(state, file) do
+    dets_state = %{time: (DateTime.utc_now |> DateTime.to_string()), state: state}
+    case :dets.insert(String.to_atom(file), {:connect, dets_state}) do
```

Comment:
The `dets` API is stateful, you need to keep a reference around to write to a file.

Reply:
Well sometimes you go with luck, turn out that the ref is the same as the filename. Thanks